components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityStoreEventListener.java

```diff
@@ -283,15 +290,64 @@ public boolean doPostGetUserClaimValues(String userName, String[] claims, String
         }
 
-        // No need to separately handle if identity `data store is user store based
-        if (isUserStoreBasedIdentityDataStore() || isStoreIdentityClaimsInUserStoreEnabled(storeManager)) {
-            return true;
-        }
-
         if (claimMap == null) {
             claimMap = new HashMap<>();
         }
 
+        // Extract claims that have a configured store location.
+        Iterator<Map.Entry<String, String>> claimMapIterator = claimMap.entrySet().iterator();
+        String storeLocation ;
+        // This map will hold the claims that have a configured store location.
+        Map<String, List<String>> storeLocationConfiguredClaims = new HashMap<>();
+
+        if(log.isDebugEnabled()){
+            log.debug("Extracting store location configured claims for user : " +
+                    userName + " in doPostGetUserClaimValues method.");
+        }
+
+        //Retrieve tenant domain from user.
+        int tenantId;
+        try{
+            tenantId = storeManager.getTenantId();
+        }catch (UserStoreException e) {
+            log.error("Error while retrieving tenant ID for user: " + userName, e);
+            return false;
+        }
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(tenantId);
+
+
+        // Iterate through the claimMap to find claims with a configured store location.
+        while (claimMapIterator.hasNext()) {
+            Map.Entry<String, String> claim = claimMapIterator.next();
+            String key = claim.getKey();
+            String value = claim.getValue();
+            if (StringUtils.isEmpty(key)) {
+                continue;
+            }
+            try {
+                storeLocation = claimMetadataManagementService.getLocalClaim(key, tenantDomain).get().getClaimProperty(ClaimConstants.SKIP_USER_STORE);
+            } catch (ClaimMetadataException e) {
+                log.error("Error while retrieving claim metadata for claim: " + key, e);
+                storeLocation = null;
+            }
+            if (!StringUtils.isEmpty(storeLocation)){
+                // If the claim has a configured storing location, we remove the claim from the claimMap
+                // and add it to the storeLocConfiguredClaims map with store location as the second value.
+                storeLocationConfiguredClaims.put(key, Arrays.asList(value, storeLocation));
+                claimMapIterator.remove();
+            }
+        }
+
+        // When user store–based identity data storage is enabled at the server level,
+        // attributes (claims) can be handled in one of two ways:
+        // 1. All claims are stored in the user store.
+        // 2. Specific claims can be configured with a storage location and stored in the identity data store.
+        if (isUserStoreBasedIdentityDataStore() || isStoreIdentityClaimsInUserStoreEnabled(storeManager)) {
+            if (storeLocationConfiguredClaims.isEmpty()){
+                return true;
+            }
```

Comment:
We can combine this to single if condition right?

```suggestion
        if (isUserStoreBasedIdentityDataStore() || isStoreIdentityClaimsInUserStoreEnabled(storeManager) || storeLocationConfiguredClaims.isEmpty()){
                return true;
            }
```